proposals/0023-asynchronous-program-execution-and-broadcast.md

```diff
@@ -0,0 +1,371 @@
+
+* Leader - the current leader for the slot that will build a PoH
+ledger full of Votes and UserBlockEntries.
+
+* Builder - a node that is scheduled to build a block with non
+vote transactions
+
+* VoteHash - like BankHash, but for vote only transactions.
+
+* BundleTransactions - a transaction signed by the builder that can
+list transactions in the builders UserBlock to be executed in
+one ordered batch. This transaction can also add a priority fee
+such that the entire bundle is prioritized for execution as a batch.
+This transaction can only appear in the builders UserBlock and can
+only reference transactions prior to it in the UserBlock. TBD on
+format.
+
+## Detailed Design
+
+### Overview
+
+Leaders are scheduled to build blocks as they are currently by
+the LeaderSchedule.
+
+Builder's are scheduled along side leaders to build UserBlocks -
+blocks of non vote transactions. N number of builders can be scheduled
+concurrently to build blocks.
+
+While a leader is scheduled, they receive and encode votes as normal.
+Any well formed UserBlocks that were received from the previous or
+current UserBlockSlot are added to the leaders PoH ledger as
+UserBlockEntry.
+
+The N concurrent builders create blocks out of user transactions.
+These are transmitted to the cluster via turbine concurrently with
+the scheduled leader and other builders.  The leader receives and
+decodes them and generates a UserBlockEntry, and adds it to PoH as
+soon as the leaders PoH has started the UserBlockSlot.
+
+Validators can vote on leader blocks by executing the votes, but
+before completing the execution of UserBlock transactions.
+
+### Fork Choice
+
+If a validator doesn't have the builder's UserBlock, the validator
+doesn't vote on the proposed block and tries to repair the UserBlock.
+That fork is ignored for fork choice.
+
+Otherwise the validator evaluates the forks as usual.
+
+### UserBlock builder schedule
+
+Each block is sloted to support N user blocks. Some of these builders
+are randomly scheduled, some of these are persistent. For example,
+there are 10 UserBlock builders per network block, that means 10
+builders are each assigned 10% of the shreds and 10% of the compute
+available to the block.
+
+#### Randomized UserBlock builders
+
+Randomized UserBlock builders are assigned at the same time as the
+leader schedule is created based on stake weighted leader distribution.
+
+The benefit to more than 1 random UserBlock builder per leader
+block is that they are likely geographically distributed and users
+will be able to pick the closest one.
+
+The downside is that resource management becomes more complex. Each
+builder has 1/N compute and shred capacity and users have no idea
+which one is saturated when sending their transaction.  It's likely
+that priority fee floor will be different at each UserBlock builder.
+
+#### Persistent UserBlock builders
+
+Persistent UserBlock slots for the epoch are auctioned of to the top
```

Comment:
how do you prevent the same N builders from winning every auction. imagine a whale builder bidding at a small loss every epoch pricing everyone out, only to earn all the mev for a volatile epoch

Reply:
Someone else can outbid them. As long as the auction process isn't censored it's fine.  That's kind of the point of persistent builders.

Reply:
This is kinda the same with flashbots right now with 3 block builders making [70% of the](https://www.relayscan.io/) blocks because validators are looking for blocks with higher gas fees and optimized [uncle rate](https://blog.ethereum.org/2016/10/31/uncle-rate-transaction-fee-analysis)  and don't care much where the block is coming from.